proxy/coco/src/request/states.rs

```diff
@@ -0,0 +1,340 @@
+//! The state types and traits that we can use for [`super::Request`]'s `S` parameter.
+
+use std::{
+    collections::HashMap,
+    fmt,
+    ops::{Add, AddAssign, Deref},
+};
+
+use serde::{Deserialize, Serialize};
+
+use librad::{peer::PeerId, uri::RadUrn};
+
+use super::sealed;
+
+impl sealed::Sealed for Created {}
+impl sealed::Sealed for Requested {}
+impl sealed::Sealed for Found {}
+impl sealed::Sealed for Cloning {}
+impl sealed::Sealed for Cancelled {}
+
+// State Types
+
+/// An enumeration of the different states a `Request` can be in. This is useful if we want to
+/// convey the state information without any of the other state data.
+#[derive(Clone, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
+pub enum RequestKind {
+    /// The initial state where the `Request` has been created.
+    Created,
+
+    /// The state where the `Request` has been requested.
+    Requested,
+
+    /// The state where the `Request` has found at least one peer.
+    Found,
+
+    /// The state where the `Request` has is cloning from a peer.
+    Cloning,
+
+    /// The state where the `Request` has successfully cloned from a peer.
+    Cloned,
+
+    /// The state where the `Request` has been cancelled.
+    Cancelled,
+
+    /// The state where the `Request` has timed out.
+    TimedOut,
+}
+
+impl fmt::Display for RequestKind {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{:?}", self)
+    }
+}
+
+/// The initial state for a `Request`. It has simply been created.
+#[derive(Clone, Debug, Hash, PartialEq, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+/// The initial state for a `Request`. It has simply been created.
+pub struct Created {}
+
+/// The state signifying that the `Request` has been kicked-off.
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct Requested {
+    /// A set of found peers and the lifecycle of clone attempts made on those peers.
+    pub(super) peers: HashMap<PeerId, Status>,
+}
+
+/// `Status` represents the lifecycle of a clone attempt, when paired with a `PeerId`.
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub enum Status {
```

Comment:
```suggestion
pub enum CloneStatus {
```

The namespace doesn't specify that it's related to "Clone"ing so I'd prefer to see it in the type.

Reply:
It's also not necessarily specific to cloning. This data could represent any kind of process that has three states. I suppose the only weird thing is that it doesn't have a `Done` since that's handled by a Request's state transition.

That is to say, I understand where the suggestion is coming from but disagree that it should be done.

Reply:
> It's also not necessarily specific to cloning

Hm, the docs suggest otherwise:

> `Status` represents the lifecycle of a clone attempt

Reply:
touché! I shall change the docs